### ISSUE TITLE
[Request] Wayland/Wlroots support?

### ISSUE THREAD

**Issue by HolyStephano** [phuhl/linux_notification_center#26]
I love the look of this notification daemon, and notice it has GTK3 in the dependencies. However, I'm on sway full time right now, and there's no good way for this to work on sway without implementing a layer-shell surface with wlroots.

**Comment by phuhl** [comment]
Hi @HolyStephano, 

What exactly would have to be done for making it run on wayland? GTK3 seems to be supported by wayland. Sadly, I am not yet familiar with writing wayland applications.

**Comment by HolyStephano** [comment]
I personally don't, but my friend has experience (he's the developer of Wofi)  with wlroots, I'm sure he could help you out. He doesn't use github, is there any IRC channels or something to reach you in regularly?

**Comment by phuhl** [comment]
git@ph-uhl.com would be best, thx

**Comment by HolyStephano** [comment]
Awesome! he should be emailing you soon, make sure to check your spam folder if it runs through gmail at all.

**Comment by Garionion** [comment]
I'm also on sway, and it basically works, the notifications  are just not positionable. they are right in the center of the screen. started from shell, there is just this message:
`Gdk-Message: 16:46:41.010: Window 0x428ce10 is a temporary window without parent, application will not be able to position it on screen.`

and also I can't get rid of the notifications reliable by hand (they disapear after some time and some notifications I can dismiss by clicking on them, but not all notifications react like that)

**Comment by travankor** [comment]
To have positionable notifications on sway/wlroots, you will need to use [gtk-layer-shell](https://github.com/wmww/gtk-layer-shell).

**Comment by phuhl** [comment]
Hey @HolyStephano, in #99 someone uses the DNC on Sway, does it work like that for you, too? Then we could document the process and close this issue

**Comment by HolyStephano** [comment]
The way they're using it is by setting GDK_BACKEND=x11 which forces it to use xwayland. While this would work, it isn't "correct" for wlroots compositors, and just tends to either look bad, break, or not function at all. Actual wlroots support would be great. There's a [gtk-layer-shell](https://github.com/wmww/gtk-layer-shell) project that you could look into, or the actual [wlroots protocol itself](https://github.com/swaywm/wlroots)

**Comment by phuhl** [comment]
Sadly I lack in capacity to do this, currently. Also I have little to no experience with wayland, so I won't implement this for now. But if someone gets it to work and opens a working PR, I will merge it.

**Comment by elkowar** [comment]
Quick note on this: If you're planning to migrate to gtk4 (#136), you're effectively giving up on this (and vice versa). Gtk4 doesn't do layer-shell at all, not even with the gtk-layer-shell "hack", so,.... good luck ^^'

**Comment by phuhl** [comment]
I thought, GTK4 was design with wayland in mind? Anyways, I doubt, there will be a merge of #136 soon as GTK4 has no way to move windows, which is absolutely something we need. Actually, the GTK4 reasoning behind this is, a) moving windows should be done by the window manager, b) they can't move windows across platforms, because it's not possible in wayland (like, wtf). I noticed that after I was nearly done with the transition. There probably are ways to get arround it, but it is not clear if the benefits justify the effort.

**Comment by elkowar** [comment]
Doing window movement in X manually is mostly trivial (i'm working on something very simmilar, and am in the process of migrating to gtk4 currently), although still anoying.
On Wayland they completely removed the code required for things like gtk-layer-shell to work. This means that gtk windows can only render to application surfaces, which, under wayland, have no support for client-controlled window position - that's what layershell is there for, which isn't supported.

you _could_ position windows on wayland before (gtk-layer-shell was the thing to use) but gtk4 removed that possibility, pretty much without a chance of bringing it back.

**Comment by phuhl** [comment]
> i'm working on something very simmilar, and am in the process of migrating to gtk4 currently

I assume, you are talking about EWW?

How did you solve the issue? No wayland support?

Could you point me at where the code for moving windows on X is in your project? I'd add a reference to #136 in case I come arround and finish the migration.

**Comment by elkowar** [comment]
Yea, talking about eww.
My wayland """solution""" for now is to not support the window-positioning side of things in wayland, but to instead just do nothing there. Then, I'll try to write some documentation about how to configure window position via rules in some of the more popular wayland compositors - doing that should be possible in most compositors.
As the rendering still works, wayland is still """supported""", just missing some key things that will need to be replicated on the compositor side.

The relevant code I wrote to replace the removed GTK calls is pretty much all in https://github.com/elkowar/eww/blob/gtk4/src/display_backend/x11.rs, being used in https://github.com/elkowar/eww/blob/gtk4/src/app.rs
Still imperfect, but I've managed to replicate the old gtk functionality moooooostly, with a tiny bit of jankyness in some rare edgecases.

**Comment by 0reoKing** [comment]
> Quick note on this: If you're planning to migrate to gtk4 (#136), you're effectively giving up on this (and vice versa). Gtk4 doesn't do layer-shell at all, not even with the gtk-layer-shell "hack", so,.... good luck ^^'

https://github.com/wmww/gtk-layer-shell/issues/37
`I've started work on option 3 (porting without upstream changes). Looks like it's probably going to be possible. Initial proof-of-concept works. Code is messy, incomplete, and there's no popup support yet. WIP branch: gtk4.`


Looks like layer shell support for gtk4 is in progress

**Comment by godalming123** [comment]
Wayland support would defiantly be a nice feature since there isn't any good notification managers on wayland with a notifications overview unless you want to make one from the ground up which doesn't sound fun.

**Comment by akhiljalagam** [comment]
for those who use sway: https://github.com/ErikReider/SwayNotificationCenter

**Comment by phuhl** [comment]
A first version of wayland support now exists in #242. Feel free to test it

**Comment by godalming123** [comment]
> for those who use sway: https://github.com/ErikReider/SwayNotificationCenter

This actually supports more than just sway, it supports everything that has `wlr_layer_shell` (so most things that are wlroots based like hyprland)

**Comment by wineee** [comment]
> Looks like layer shell support for gtk4 is in progress

https://github.com/wmww/gtk4-layer-shell gtk4 support has been implemented

**Comment by phuhl** [comment]
@wineee Yes, I saw that, too. Now we need the Haskell bindings for GTK4 Layer Shell and also a solution for moving windows with GTK 4 on X Server as I don't want to brake compatibility.

**Comment by leon-erd** [comment]
Would love to see support for wlroots! :)
